lxd/storage_lvm.go

```diff
@@ -1562,25 +1666,308 @@ func (s *storageLvm) ContainerSnapshotCreateEmpty(snapshotContainer container) e
 }
 
 func (s *storageLvm) ContainerBackupCreate(backup backup, sourceContainer container) error {
+	logger.Debugf("Creating LVM storage volume for backup \"%s\" on storage pool \"%s\".", s.volume.Name, s.pool.Name)
+
+	// mount storage
+	ourStart, err := sourceContainer.StorageStart()
+	if err != nil {
+		return err
+	}
+	if ourStart {
+		defer sourceContainer.StorageStop()
+	}
+
+	// Create the path for the backup.
+	baseMntPoint := getBackupMountPoint(s.pool.Name, backup.Name())
+	targetBackupContainerMntPoint := fmt.Sprintf("%s/container", baseMntPoint)
+	err = os.MkdirAll(targetBackupContainerMntPoint, 0711)
+	if err != nil {
+		return err
+	}
+
+	snapshots, err := sourceContainer.Snapshots()
+	if err != nil {
+		return err
+	}
+
+	rsync := func(oldPath string, newPath string, bwlimit string) error {
+		output, err := rsyncLocalCopy(oldPath, newPath, bwlimit)
+		if err != nil {
+			s.ContainerBackupDelete(backup.Name())
+			return fmt.Errorf("failed to rsync: %s: %s", string(output), err)
+		}
+		return nil
+	}
+
+	bwlimit := s.pool.Config["rsync.bwlimit"]
+	if !backup.containerOnly && len(snapshots) > 0 {
+		// /var/lib/lxd/storage-pools/<pool>/backups/<container>/snapshots
+		targetBackupSnapshotsMntPoint := fmt.Sprintf("%s/snapshots", baseMntPoint)
+		err = os.MkdirAll(targetBackupSnapshotsMntPoint, 0711)
+		if err != nil {
+			return err
+		}
+
+		for _, snap := range snapshots {
+			_, err := s.ContainerSnapshotStart(snap)
+			if err != nil {
+				return err
+			}
+
+			snapshotMntPoint := getSnapshotMountPoint(s.pool.Name, snap.Name())
+			_, snapName, _ := containerGetParentAndSnapshotName(snap.Name())
+			target := fmt.Sprintf("%s/%s", targetBackupSnapshotsMntPoint, snapName)
+			err = rsync(snapshotMntPoint, target, bwlimit)
+			s.ContainerSnapshotStop(snap)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	snapshotSuffix := uuid.NewRandom().String()
+	sourceLvmDatasetSnapshot := fmt.Sprintf("snapshot-%s", snapshotSuffix)
+
+	// /var/lib/lxd/storage-pools/<pool>/containers/<container>
+	tmpContainerMntPoint := getContainerMountPoint(s.pool.Name, sourceLvmDatasetSnapshot)
+	err = os.MkdirAll(tmpContainerMntPoint, 0711)
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpContainerMntPoint)
+
+	_, err = s.createSnapshotLV(s.pool.Name, sourceContainer.Name(),
+		storagePoolVolumeAPIEndpointContainers, containerNameToLVName(sourceLvmDatasetSnapshot),
+		storagePoolVolumeAPIEndpointContainers, false, s.useThinpool)
+	if err != nil {
+		return err
+	}
+	defer removeLV(s.pool.Name, storagePoolVolumeAPIEndpointContainers,
+		containerNameToLVName(sourceLvmDatasetSnapshot))
+
+	if !shared.IsMountPoint(tmpContainerMntPoint) {
```

Comment:
This needs to go away:
1. `doContainerMount()` will already check that
2. you're not tacking the necessary lock